partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,917 @@
+                            "No Typing Information for {:?} in closed Schema {:?}",
+                            &key, &derived_type
+                        )));
+                        None
+                    }
+                }
+            }
+        } else if derived_type.is_any() {
+            Some(any!())
+        } else {
+            self.errors.push(TypingError::IllegalState(format!(
+                "Illegal Derive Type {:?}",
+                &derived_type
+            )));
+            None
+        }
+    }
+
+    fn derived_type(&mut self, ty_ctx: &TypeEnvContext) -> PartiqlType {
+        let ty = ty_ctx.derived_type();
+        ty.clone()
+    }
+
+    fn local_type_ctx(&mut self) -> TypeEnvContext {
+        let out = self
+            .type_env_stack
+            .last()
+            .ok_or_else(|| TypingError::IllegalState("Malformed TypeEnv stack".to_string()));
+
+        match out {
+            Ok(out) => out.clone(),
+            Err(err) => {
+                self.errors.push(err);
+                TypeEnvContext::new()
+            }
+        }
+    }
+
+    fn lookup_order(&self, v: &ValueExpr) -> LookupOrder {
+        match v {
+            ValueExpr::VarRef(_, varef_type) => match varef_type {
+                VarRefType::Global => GlobalLocal,
+                VarRefType::Local => LocalGlobal,
+            },
+            _ => match self.current_bindings_op {
+                Some(BindingsOp::Scan(_)) => GlobalLocal,
+                _ => LocalGlobal,
+            },
+        }
+    }
+
+    fn resolve_global_then_local(&mut self, key: &SymbolPrimitive) -> PartiqlType {
+        let ty = self.resolve_global(key);
+        match ty.is_undefined() {
+            true => self.resolve_local(key),
+            false => ty,
+        }
+    }
+
+    fn resolve_local_then_global(&mut self, key: &SymbolPrimitive) -> PartiqlType {
+        let ty = self.resolve_local(key);
+        match ty.is_undefined() {
+            true => self.resolve_global(key),
+            false => ty,
+        }
+    }
+
+    fn resolve_global(&mut self, key: &SymbolPrimitive) -> PartiqlType {
+        if let Some(type_entry) = self.catalog.resolve_type(key.value.as_str()) {
+            let ty = self.element_type(type_entry.ty());
+            ty
+        } else {
+            undefined!()
+        }
+    }
+
+    fn resolve_local(&mut self, key: &SymbolPrimitive) -> PartiqlType {
+        for type_ctx in self.type_env_stack.iter().rev() {
+            if let Some(ty) = type_ctx.env().get(key) {
+                return ty.clone();
+            }
+        }
+
+        undefined!()
+    }
+
+    fn type_with_undefined(&mut self, key: &SymbolPrimitive) {
+        if let TypingMode::Permissive = &self.typing_mode {
+            let type_ctx = ty_ctx![(&ty_env![(key.clone(), missing!())], &undefined!())];
+
+            self.type_env_stack.push(type_ctx);
+        }
+    }
+
+    fn get_singleton_type_from_env(&mut self) -> PartiqlType {
```

Comment:
Sorry think I missed this from a prior review. In what contexts should this `fn` be called? And why does it expect just one type in the typing environment (is it related to duplicates at all?)? Seems like it's currently called when extracting the final output schema from the typed plan and when creating the projection list schema.

Reply:
It's not related to duplicates, it is a helper function to extract one type out of the environment when we expect it. As far as I can tell today, this is mainly in projections, when we expect to infer one type from the project list items.